File: day08/puzzle.py
from aoc import AocPuzzle


TEST_ANSWERS = {
    "part1": 26,
    "part2": 61229
}

UNIQUE_SEGMENT_MAP = {
    2: 1,
    3: 7,
    4: 4,
    7: 8
}


def is_subset(subset: str, superset: str) -> bool:
    for char in subset:
        if not char in superset:
            return False

    return True


class Entry:
    def __init__(self, entry: str):
        self.patterns, self.output = [part.split(" ") for part in entry.strip().split(" | ")]


class Puzzle(AocPuzzle[list[Entry], int]):
    def __init__(self):
        super().__init__(TEST_ANSWERS)


    def parse_input(self, lines: list[str]):
        return [Entry(line) for line in lines]


    def run_part1(self):
        return sum([len([digit for entry in self._input for digit in entry.output if len(digit) in UNIQUE_SEGMENT_MAP])])


    def run_part2(self):
        values: list[int] = []
        for entry in self._input:
            segment_map_5: list[str] = []
            segment_map_6: list[str] = []
            number_map: dict[int, str] = {}
            for pattern in entry.patterns:
                length = len(pattern)
                if length == 5:
                    segment_map_5.append(pattern)
                elif length == 6:
                    segment_map_6.append(pattern)
                else:
                    number_map[UNIQUE_SEGMENT_MAP[length]] = pattern

            for i in range(3):
                if is_subset(number_map[1], segment_map_5[i]):
                    number_map[3] = segment_map_5.pop(i)
                    break

            for i in range(3):
                if is_subset(number_map[4], segment_map_6[i]):
                    number_map[9] = segment_map_6.pop(i)
                    break

            if is_subset(number_map[1], segment_map_6[0]):
                number_map[0] = segment_map_6[0]
                number_map[6] = segment_map_6[1]
            else:
                number_map[6] = segment_map_6[0]
                number_map[0] = segment_map_6[1]

            if is_subset(segment_map_5[0], number_map[6]):
                number_map[5] = segment_map_5[0]
                number_map[2] = segment_map_5[1]
            else:
                number_map[2] = segment_map_5[0]
                number_map[5] = segment_map_5[1]

            output_map = {"".join(sorted(v)): k for k, v in number_map.items()}
            values.append(sum([10 ** i * output_map["".join(sorted(entry.output[-i - 1]))] for i in range(4)]))

        return sum(values)

File: day20/puzzle.js
import { createAocPuzzle } from "./aoc.js";

const testAnswers = {
  part1: 35,
  part2: 3351
};

function parseInput(lines) {
  const enhancement = lines[0].split("").map(char => +(char == "#"));

  const width = lines[2].length;
  const image = [new Int8Array(width + 4), new Int8Array(width + 4)];
  for (let i = 2; i < lines.length; ++i) {
    const row = new Int8Array(width + 4);
    for (let j = 0; j < width; ++j) {
      row[j + 2] = +(lines[i][j] == "#");
    }
    image.push(row);
  }
  image.push(new Int8Array(width + 4), new Int8Array(width + 4));

  return { enhancement, image };
}

function enhanceImage(image, enhancement) {
  const width = image[0].length;
  const height = image.length;
  const fill = (image[0][0] ? enhancement[enhancement.length - 1] : enhancement[0]);
  const enhancedImage = [new Int8Array(width + 4).fill(fill), new Int8Array(width + 4).fill(fill)];
  for (let i = 0; i < height; ++i) {
    enhancedImage.push(new Int8Array(width + 4).fill(fill));
  }

  for (let y = 1; y < height - 1; ++y) {
    for (let x = 1; x < width - 1; ++x) {
      const index = (
        256 * image[y - 1][x - 1] +
        128 * image[y - 1][x] +
        64 * image[y - 1][x + 1] +
        32 * image[y][x - 1] +
        16 * image[y][x] +
        8 * image[y][x + 1] +
        4 * image[y + 1][x - 1] +
        2 * image[y + 1][x] +
        image[y + 1][x + 1]
      );
      enhancedImage[y + 2][x + 2] = enhancement[index];
    }
  }

  enhancedImage.push(new Int8Array(width + 4).fill(fill), new Int8Array(width + 4).fill(fill));
  return enhancedImage;
}

function countPixels(image) {
  let count = 0;
  for (let y = 2; y < image.length - 2; ++y) {
    for (let x = 2; x < image[0].length - 2; ++x) {
      count += image[y][x];
    }
  }
  return count;
}

function runPart1(input) {
  let image = enhanceImage(input.image, input.enhancement);
  image = enhanceImage(image, input.enhancement);
  return countPixels(image);
}

function runPart2(input) {
  let image = input.image;
  for (let i = 0; i < 50; ++i) {
    image = enhanceImage(image, input.enhancement);
  }
  return countPixels(image);
}

export function createPuzzle() {
  return createAocPuzzle({ parseInput, runPart1, runPart2, testAnswers });
}
